src/utils/WebSocketsService.ts

```diff
@@ -69,3 +74,106 @@ export default {
   isConnected,
   send,
 };
+
+/**
+ * This hook exposes SocketWrapper containing an instance of the created WebSocket object
+ * and the send function. `send` function "wraps" WebSocket's `send` to add
+ * checks on the socket object's ready state before performing actual send.
+ *
+ * @param props { messageHandler, userAuthToken }
+ */
+export function useWebSocket(props) {
+  const { messageHandler, userAuthToken } = props || {};
+
+  /**
+   * Main instance of websocket which events and states are tracked.
+   * This is stored in the instance property of the SocketWrapper - the object that
+   * is passed as a value to the SocketContext defined in the App.tsx
+   */
+  const [instance, setInstance] = useState<WebSocket>(null);
+
+  const [readyState, setReadyState] = useState(undefined);
+
+  const sendFn = (i: WebSocket, token: string) => {
+    return message => {
+      if (i) {
+        i.send(JSON.stringify({ ...message, token }));
+      }
+    };
+  };
+
+  /**
+   * socketWrapper state stores the current instance and the `send` function.
+   * `send` function is run the MiddleWare executes `sendMessage()` defined in
+   * a WebSocket-supported action creator (ie, boardActions#openedBoard)
+   */
+  const [socketWrapper, setSocketWrapper] = useState<ISocketWrapper>({
+    instance,
+    send: sendFn(instance, userAuthToken),
+  });
+
+  useEffect(
+    () => {
+      if (!instance) {
+        const ws = new WebSocket(`${config.WS_URL}/ws`);
+        ws.binaryType = 'arraybuffer';
+
+        ws.onopen = function() {
+          setReadyState(ws.readyState);
+        };
+
+        ws.onmessage = event => {
+          if (messageHandler) {
+            messageHandler(event);
+          }
+        };
+
+        ws.onclose = function() {
+          setReadyState(ws.readyState);
+
+          // Try to reconnect
+          // We handle reconnection by detecting CLOSED readyState in useEffect below
+        };
+
+        setInstance(ws);
+      }
+    },
+    [instance]
+  );
+
+  /**
+   * Closes socket instance and nullifies instance's readyState is CLOSED (3)
+   * This kickoffs useEffect above
+   */
+  useEffect(
+    () => {
+      if (readyState === 3) {
+        instance.close();
+        setInstance(null);
+      }
+    },
+    [readyState]
+  );
+
+  /**
+   * Creates and sets SocketWrapper when instance and userAuthToken changes
+   */
+  useEffect(
+    () => {
+      if (instance) {
+        setSocketWrapper({
+          instance,
+          send: sendFn(instance, userAuthToken),
+        });
+      }
+    },
+    [instance, userAuthToken]
```

Comment:
The auth token is a prop, so I'm not sure if we need to do anything special when it changes, especially if we inline sendFn. In that case can we merge this useEffect with the first one?

Reply:
oh so in the initial run (when the first `useWebSocket` is run) the auth token doesn't exist yet, so I was just thinking of running this useEffect when `userAuthToken` finally gets the updated in https://github.com/srcclr/sapling-frontend/pull/32/files/e059422620544762f4e8a7c6514d2e53061478fe#diff-c8a101915dc6871f1f3d7ad2208d8655R53. I separated it to another `useEffect` thinking that it'd be good to capture that specific task separately for clarity.

Reply:
Okay, might be worth leaving a comment on why we separate it.